h/migrations/versions/9f5e274b202c_update_all_document_web_uris.py

```diff
@@ -0,0 +1,116 @@
+"""
+Update all Document.web_uris.
+
+Revision ID: 9f5e274b202c
+Revises: e10ce4472966
+Create Date: 2017-01-20 16:07:03.442975
+"""
+
+from __future__ import unicode_literals
+
+from collections import namedtuple
+import logging
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.orm import subqueryload
+
+from memex._compat import urlparse
+
+
+revision = '9f5e274b202c'
+down_revision = 'e10ce4472966'
+
+
+Base = declarative_base()
+Session = sessionmaker()
+log = logging.getLogger(__name__)
+
+
+class Window(namedtuple('Window', ['start', 'end'])):
+    pass
+
+
+class Document(Base):
+    __tablename__ = 'document'
+    id = sa.Column(sa.Integer, primary_key=True)
+    updated = sa.Column(sa.DateTime)
+    web_uri = sa.Column(sa.UnicodeText())
+    document_uris = sa.orm.relationship('DocumentURI',
+                                        backref='document',
+                                        order_by='DocumentURI.updated.desc()')
+
+    def updated_web_uri(self):
+        def first_http_url(type_=None):
+            for document_uri in self.document_uris:
+                uri = document_uri.uri
+                if type_ is not None and document_uri.type != type_:
+                    continue
+                if urlparse.urlparse(uri).scheme not in ['http', 'https']:
+                    continue
+                return document_uri.uri
+
+        return (first_http_url(type_='self-claim') or
+                first_http_url(type_='rel-canonical') or
+                first_http_url())
+
+
+class DocumentURI(Base):
+    __tablename__ = 'document_uri'
+    id = sa.Column(sa.Integer, primary_key=True)
+    updated = sa.Column(sa.DateTime)
+    uri = sa.Column(sa.UnicodeText)
+    type = sa.Column(sa.UnicodeText,
+                     nullable=False,
+                     default='',
+                     server_default='')
+    document_id = sa.Column(sa.Integer,
+                            sa.ForeignKey('document.id'),
+                            nullable=False)
+
+
+def upgrade():
+    session = Session(bind=op.get_bind())
+
+    windows = _fetch_windows(session)
```

Comment:
This is totally fine for this migration, but in case you haven't seen it, we've added a more efficient way of doing windowed queries with the new [`memex.util.query.column_windows`](https://github.com/hypothesis/h/blob/master/src/memex/util/query.py#L8), and an example of how it is used is in the [`memex.search.index.BatchIndexer`](https://github.com/hypothesis/h/blob/master/src/memex/search/index.py#L158-L171).